### ISSUE TITLE
Rounding issues in TransformState

### ISSUE THREAD

**Issue by brunoabinader** [mapbox/mapbox-gl-native#12420]
In `TransformState`, some getters/setters use double floating point, and some others use single floating point precision. 

For instance, [TransformState::getFieldOfView()](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/map/transform_state.hpp#L67) returns a `float`, but [fov](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/map/transform_state.hpp#L136) is stored as `double`. See example below:
```c++
#include <iostream>

int main()
{
    double fov = 0.6435011087932844;
    double fov1 = float(fov);
    printf("%.25lf\n", fov);
    printf("%.25lf\n", fov1);
    return 0;
}
```
The example code above produces the following output:
```
0.6435011087932843709680242
0.6435011029243469238281250
```
A `double` that is casted into a `float` and then [used as a double again](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/map/transform_state.cpp#L39) is prone to rounding errors e.g. `tileCover()` will generate two tiles: 0/0/0 and 0/1/0 instead of one for the following utests:
```
[  FAILED  ] Source.RasterTileFail
[  FAILED  ] Source.RasterDEMTileFail
[  FAILED  ] Source.VectorTileFail
[  FAILED  ] Source.RasterTileCorrupt
[  FAILED  ] Source.RasterDEMTileCorrupt
[  FAILED  ] Source.VectorTileCorrupt
...
      Expected: OverscaledTileID(0, 0, 0)
      Which is: 0/0/0=>0+0
To be equal to: tileID
      Which is: 0/0/0=>0+1
```

/cc @mapbox/gl-core

**Comment by ChrisLoer** [comment]
Good catch and good to keep in mind! How would you code defensively against this kind of problem? We do a lot of switching between `double` and `float` (e.g. `Feature = mapbox::geometry::feature<double>`, but symbol layout/projection and collision detection generally use `float`). I don't know that we really have strong reasons for using one or the other, but in large data structures/tight loops, I can see the argument that `float` is half the size and has all the precision we should need...
